create how to ask discourse questions (#36)
pymc-devs/pymc.io: community/how_to_ask_on_discourse.md

(how_to_ask_on_discourse)=
# Asking Questions on the PyMC Discourse

This document provides some advice for those asking questions on the [PyMC Discourse site](https://discourse.pymc.io/), the main channel for communication among the PyMC community.  If you have questions regarding modeling, PyMC/Aesara/arviz usage, or Bayesian modeling in general, please create a new post and someone is likely help you out.

The items in this guide are designed to help maximize the chances that you good answers quickly.

> - {ref}`search`

Reviewer: In general I'd say the toc already appears in the right sidebar automatically, but I think it is good to have it here as inline content. 

I would update the targets to be `ask_discourse/search`... or something like this to avoid potential clashes. In the case of the `search` target this is a must as it is a base sphinx protected target pointing to the search page, for the other ones it is not necessary but strongly recommended

- {ref}`problem`
- {ref}`mwe`
- {ref}`title`

> - {ref}`version`

Reviewer: ```suggestion
- {ref}`versions`
```

- {ref}`tags`
- {ref}`example`

(search)=
## Search

> The PyMC Discourse is full of questions that have already received useful answers.  So definitely use the search function to see if anyone has asked questions that are similar to yours.  Try different combinations of keywords because past questions don't necessarily include a great {ref}`title`.  If you find similar questions, but the answers do not solve your problem, include links to these past questions in your topic and explain why the answers do not solve your problem.  You can also search for topics by selecting relevant {ref}`tags` (or filter search results by tag).

Reviewer: I think here we should use custom text for the title reference instead of having sphinx fill it with the actual title as it breaks the reading flow. The section title being referenced is "A good title" so it now reads `a great A good title.`

```suggestion
The PyMC Discourse is full of questions that have already received useful answers.  So definitely use the search function to see if anyone has asked questions that are similar to yours.  Try different combinations of keywords because past questions don't necessarily include a great {ref}`title <title>`.  If you find similar questions, but the answers do not solve your problem, include links to these past questions in your topic and explain why the answers do not solve your problem.  You can also search for topics by selecting relevant {ref}`tags` (or filter search results by tag).
```

> 

Reviewer: Not sure if here or below in the problem section. Also mention to include the links to the pages in the docs they have read/adapted the code from.

Author: What's an easy way to get a sphinx reference to the stable version of the docs?

(mwe)=
## Minimum Working Example
Please include an example to demonstrate the problem you are facing.  There are several critical properties of your example:

**Minimal**

Do not paste your entire script.  Only include the parts of your code/model/data/processing that are necessary to demonstrate your problem.  Do not include unnecessary import statements, unused function definitions, or code that will not be executed.

If you are currently running your code in a Jupyter notebook, export it as an executable Python script or reproduce the relevant code in a new Python script.

**Self-contained**

Anyone should be able to immediately run your example code.  This means that your example should include all import statements.  Your example should include any custom functions that are necessary.  If you code relies on data that is loaded from an external file/site, consider creating or generating toy data in your example itself:
  - `my_data = [[2, 7], [31, 9]]`
  - `mydata = np.random.default_rng(12345).random((2,2))`

If there are calls that (implicitly) rely on random seeds, explicitly set the seed

>   - [`np.random.default_rng(12345)`](https://numpy.org/doc/stable/reference/random/generator.html)

Reviewer: ```suggestion
  - {func}`np.random.default_rng(12345) <numpy.random.default_rng>`
```

Highlight what you believe to be the critical aspects of your question.  Finally, if you have trouble summarizing what aspects of your problem are important, you can try waiting until you have written your entire topic, including the {ref}`mwe` before trying to write the title.

**Good titles**:
- "Why doesn't pm.sample() use all my cores?"
- "Multiprocessing BrokenPipeError: [Errno 32] Broken pipe"
- "How I get az.plot_dist() to show priors?"

**Not so good titles**:
- "How do I do Bayesian inference?"
- "Is there a better way to implement this model?"
- "Hierarchical models"

(versions)=
## Versions
Please include information about what versions of the relevant packages you have installed and how they were installed, including:

- PyMC version
- Aesara version
- Python version
- System, etc. (e.g., linux, Mac, Windows, jupyter, docker, Google Colab)
- How you installed PyMC (e.g., pip, conda, mamba)

> ## Tags <a id='tags'></a>

Reviewer: we should use only myst targets, not raw html ones

> Discourse includes two ways to classify topics: categories and tags.  Each topic belongs to exactly one category.  The most popular category is `Questions`, but `Development` and `Sharing` are also quite popular.  Each topic may have more than one tag.  When you post your topic, search through the tags for any key words that might be relevant.  Tags include things like `bug`, `gaussian_process`, and `gpu`.

Reviewer: I would highlight the importance of categories and tags even more. I think it might we worth mentioning for example that unlike GitHub, people can subscribe to specific categories or tags. 

I for example check out discourse from time to time whenever I feel like it, there is no routine, sometimes I check it every day for a while and sometimes I spend a week or more without answering. But I am subscribed to the "arviz" tag, so whenever a new question is posted using this tag I get an email notification, and I do check my email roughly every day. If the question is long or complicated I will bookmark it to answer next time I log in to discourse and have time, but if it is short, chances are I'll answer much faster. If no tag is used, I'll probably get to it next time I log in, but if it has been a longish hiatus of several days, it is possible I miss the topic altogether.

Side note: users can no longer post on "Questions" category, they need to use one of the 3 subcategories, v3, v4 or version-agnostic.

Author: Are you thinking needs to be added (or emphasized?)?  It already talks about how people can subscribe to tags and categories and it says that this is the reason that selecting categories/tags is important for getting eyeballs, etc.  Should we just try to emphasize the importance of this step more?

Reviewer: Maybe we could put the 2nd paragraph in an [fancy style box](https://jupyterbook.org/en/stable/reference/cheatsheet.html#admonitions).

Or add a "Summary" box at the top/bottom (maybe replacing the index if at the top) with bullet points summarizing the advise on the doc, for example (order now is completely random):

* Avoid the "General" category and use {ref}tags
* Use a clear {ref}title
* If code related, make sure to include a {ref}mwe
* ...

Sorry I am not being very actionable in my comments, I'd like this to pop more because I think it is key for discourse to work properly and especially to scale to more users but I'm not sure how to get there

Author: I am a bit wary of adding a "summary" at the beginning because it may discourage people from reading anything else.  But I added it at the end.  See what you think.

Categories and tags are an important part of how discourse functions.  Users can subscribe to categories and/or tags that are of interest.  This means that careful selection of a category and one or more tags can alert the most relevant users to your question and get you answers quicker.

(example)=

> ### An Example

Reviewer: I think this should be level 2 heading. And the one below level 3

(how_to_ask_on_discourse)=
# Asking Questions on the PyMC Discourse

This document provides some advice for those asking questions on the [PyMC Discourse site](https://discourse.pymc.io/), the main channel for communication among the PyMC community.  If you have questions regarding modeling, PyMC/Aesara/arviz usage, or Bayesian modeling in general, please create a new post and someone is likely help you out.

The items in this guide are designed to help maximize the chances that you good answers quickly.

- {ref}`ask_discourse/search`
- {ref}`ask_discourse/problem`
- {ref}`ask_discourse/mwe`
- {ref}`ask_discourse/title`
- {ref}`ask_discourse/versions`
- {ref}`ask_discourse/tags`
- {ref}`ask_discourse/example`

> - {ref}`ask_discourse/summary`

Reviewer: ```suggestion
- {ref}`Summary <ask_discourse/summary>`
```

(ask_discourse/search)=
## Search

> The PyMC Discourse is full of questions that have already received useful answers.  So definitely use the search function to see if anyone has asked questions that are similar to yours.  Try different combinations of keywords because past questions don't necessarily include {ref}`title`<a great title>.  You can also search disourse for topics by selecting relevant {ref}`tags` (or filter search results by tag).  You should also try searching {ref}`the PyMC documentation <pymc/>` because it covers both the PyMC API and common use-cases.  If you find something that is related but does not solve your specific problem, include links to these related pages and explain why the what you have found does not solve your problem.

Reviewer: ```suggestion
The PyMC Discourse is full of questions that have already received useful answers.  So definitely use the search function to see if anyone has asked questions that are similar to yours.  Try different combinations of keywords because past questions don't necessarily include {ref}`a great title <title>`.  You can also search discourse for topics by selecting relevant {ref}`tags` (or filter search results by tag).  You should also try searching {doc}`the PyMC documentation <pymc:index>` because it covers both the PyMC API and common use-cases.  If you find something that is related but does not solve your specific problem, include links to these related pages and explain why the what you have found does not solve your problem.
```

Highlight what you believe to be the critical aspects of your question.  Finally, if you have trouble summarizing what aspects of your problem are important, you can try waiting until you have written your entire topic, including the {ref}`mwe` before trying to write the title.

**Good titles**:
- "Why doesn't pm.sample() use all my cores?"
- "Multiprocessing BrokenPipeError: [Errno 32] Broken pipe"
- "How I get az.plot_dist() to show priors?"

**Not so good titles**:
- "How do I do Bayesian inference?"
- "Is there a better way to implement this model?"
- "Hierarchical models"

(ask_discourse/versions)=
## Versions
Please include information about what versions of the relevant packages you have installed and how they were installed, including:

- PyMC version
- Aesara version
- Python version
- System, etc. (e.g., linux, Mac, Windows, jupyter, docker, Google Colab)
- How you installed PyMC (e.g., pip, conda, mamba)

> (ask_discourse/tags)

Reviewer: ```suggestion
(ask_discourse/tags)=
```

## Tags
Discourse includes two ways to classify topics: categories and tags.  Each topic belongs to exactly one category.  The most popular categories are for questions (e.g., `Questions/v4`, `Questions/v3`), but `Development` and `Sharing` are also quite popular.  Each topic may have more than one tag.  When you post your topic, search through the tags for any key words that might be relevant.  Tags include things like `bug`, `gaussian_process`, and `gpu`.

Categories and tags are an important part of how discourse functions.  Users can subscribe to categories and/or tags that are of interest.  This means that careful selection of a category and one or more tags can alert the most relevant users to your question and get you answers quicker.

> ```{important} Categories and tags are a critical part of how Discourse works.  Many users subscribe to categories and tags that are relevant to their own expertise.  This allows them to focus exclusively on questions that they are able to answer.  So if you post a questions under a generic category (e.g., "General") or post without any tags, your question is far less likely to attract the attention of those users most capable of answer your question!```
> ```

Reviewer: ````suggestion
```{important}
Categories and tags are a critical part of how Discourse works.  Many users subscribe to categories and tags that are relevant to their own expertise.  This allows them to focus exclusively on questions that they are able to answer.  So if you post a questions under a generic category (e.g., "General") or post without any tags, your question is far less likely to attract the attention of those users most capable of answer your question!
```
````

### Starting point failure in Beta-Binomial model
I am new to PyMC and the world of Bayesian modeling.  I put together a simple coin-flipping model with a Binomial likelihood function and a Beta prior distribution.  When I run the model, the sampling seems to start, but then generates an error telling me that the starting point failed.    I tried changing the `alpha` and `beta` parameters in the Beta prior and I converted the Binomial to a Bernoulli, but none of those things seemed to help.   Here is a minimal example using the Bernoulli likelihood:

```python
import pymc as pm

my_data = 5 * [-1]

with pm.Model() as model:
    p = pm.Beta('p', alpha=1, beta=1)
    y = pm.Bernoulli('y', p=p, observed=my_data)
    idata = pm.sample()
```

Running this generates the following error:

```
SamplingError: Initial evaluation of model at starting point failed!
Starting values:
{'p_logodds__': array(0.83748875)}

Initial evaluation results:
{'p': -1.56, 'y': -inf}
```

Any ideas about what might the causing this?

PyMC version: 4.0.0\
Aesara version: 2.3.7\
Python version: 3.10.2\
Operating system: linux\
How you installed PyMC: conda

> 

Reviewer: ```suggestion

---
```

> (ask_discourse/summary)=
> ```{admonition} Summary

Reviewer: ````suggestion
```{admonition} Summary
:name: ask_discourse/summary

````

(how_to_ask_on_discourse)=
# Asking Questions on the PyMC Discourse

This document provides some advice for those asking questions on the [PyMC Discourse site](https://discourse.pymc.io/), the main channel for communication among the PyMC community.  If you have questions regarding modeling, PyMC/Aesara/arviz usage, or Bayesian modeling in general, please create a new post and someone is likely help you out.

The items in this guide are designed to help maximize the chances that you good answers quickly.

- {ref}`ask_discourse/search`
- {ref}`ask_discourse/problem`
- {ref}`ask_discourse/mwe`
- {ref}`ask_discourse/title`
- {ref}`ask_discourse/versions`
- {ref}`ask_discourse/tags`
- {ref}`ask_discourse/example`
- {ref}`Summary <ask_discourse/summary>`

(ask_discourse/search)=
## Search

> The PyMC Discourse is full of questions that have already received useful answers.  So definitely use the search function to see if anyone has asked questions that are similar to yours.  Try different combinations of keywords because past questions don't necessarily include {ref}`a great title <title>`.  You can also search discourse for topics by selecting relevant {ref}`tags` (or filter search results by tag).  You should also try searching {doc}`the PyMC documentation <pymc:index>` because it covers both the PyMC API and common use-cases.  If you find something that is related but does not solve your specific problem, include links to these related pages and explain why the what you have found does not solve your problem.

Reviewer: ```suggestion
The PyMC Discourse is full of questions that have already received useful answers.  So definitely use the search function to see if anyone has asked questions that are similar to yours.  Try different combinations of keywords because past questions don't necessarily include {ref}`a great title <ask_discourse/title>`.  You can also search discourse for topics by selecting relevant {ref}`ask_discourse/tags` (or filter search results by tag).  You should also try searching {doc}`the PyMC documentation <pymc:index>` because it covers both the PyMC API and common use-cases.  If you find something that is related but does not solve your specific problem, include links to these related pages and explain why the what you have found does not solve your problem.
```

  - `my_data = [[2, 7], [31, 9]]`
  - `mydata = np.random.default_rng(12345).random((2,2))`

If there are calls that (implicitly) rely on random seeds, explicitly set the seed
  - {func}`np.random.default_rng(12345) <numpy.random.default_rng>`
  - `pm.sample(random_seed=12345)`

**Relevant output**

This includes error messages, warnings, diagnostics, statistics, and/or plots.  Describe what about these results are relevant to your question.  What are you seeing that you **did not** expect?  What are you **not** seeing that you **did** expect?

**Your example should demonstrate your problem**

After streamlining your code down to just the essentials, make sure that you double-check that your example code still produces your problem.  If not, perhaps you accidentally solved your own problem!

(ask_discourse/title)=
## A Good Title
The title of your topic is critical because it's the first thing people will see when you submit your question.  If the question doesn't seem relevant or interesting or if people can't understand what your question might be about, then you may not get people to read and answer your question.

> Highlight what you believe to be the critical aspects of your question.  Finally, if you have trouble summarizing what aspects of your problem are important, you can try waiting until you have written your entire topic, including the {ref}`mwe` before trying to write the title.

Reviewer: ```suggestion
Highlight what you believe to be the critical aspects of your question.  Finally, if you have trouble summarizing what aspects of your problem are important, you can try waiting until you have written your entire topic, including the {ref}`ask_discourse/mwe` before trying to write the title.
```
